### PR TITLE
python backbone

### PR DESCRIPTION
**Pull request by swozniewski**
Develops python backbone such that current example analysis.cxx can be reproduced.
In addition, an exemplary shift is introduced. Note, that filtering with shift needs better treatment.
Channel specific treatment not yet implemented.
Some cpp modules are modified to better fit a common structure.

### CODE REVIEW

File: quantities.hxx [hunk 1]
```diff
@@ -2,41 +2,42 @@
...
 
 namespace quantities {
-auto pt(auto df, std::vector<std::string> varSet, const std::string &outputname,
-        const std::string &inputvector) {
+auto pt(auto df, std::vector<std::string> &varSet,
+        const std::string &outputname, const std::string &inputvector) {
     varSet.push_back(outputname);
     return df.Define(
         outputname,
         [](const ROOT::Math::PtEtaPhiMVector &p4) { return p4.pt(); },
```
Comment (stwunsch) [inline quantities.hxx:10]: Hier muesst ihr aufpassen, das schreibt euch ein double raus und ihr habt am Ende evtl ein 2x so grosses ntuple ;) Selbes gilt auch fuer die anderen quantities unten.

```suggestion
        [](const ROOT::Math::PtEtaPhiMVector &p4) { return (return) p4.pt(); },
```

File: code_generation/producer.py [hunk 2]
```diff
@@ -13,17 +13,26 @@ def __init__(self, call, inputs, output):
         self.output = output
 
         # keep track of variable dependencies
-        for q in self.inputs:
-            q.children.append(self.output)
+        if self.output != None:
+            for qy in self.inputs:
```
Comment (harrypuuter) [inline code_generation/producer.py:17]: `qy` should be `quantity` ?

File: code_generation/producer.py [hunk 3]
```diff
@@ -37,47 +46,222 @@ def writecall(self, config, shift=""):
...
         calls = []
         shifts = [""]
         if self.output != None:
-            shifts.extend(self.output.shifts)
+            shifts.extend(self.output[0].shifts)
         for shift in shifts:
-            for val in config[shift][self.vec_config]:
-                self.call = basecall.format_map(SafeDict({self.vec_config: val}))
+            # check that all config lists (and output if applicable) have same length
+            l = len(config[shift][self.vec_configs[0]])
```
Comment (harrypuuter) [inline code_generation/producer.py:72]: switch from `l` to some longer more descriptive variable name

File: code_generation/producer.py [hunk 4]
```diff
@@ -37,47 +46,222 @@ def writecall(self, config, shift=""):
...
+    PG_count = 1  # counter for internal quantities used by ProducerGroups
+
+    def __init__(self, call, inputs, output, subproducers):
+        self.call = call
+        self.inputs = inputs
+        self.output = output
+        self.producers = subproducers
+        # If call is provided, this is supposed to consume output of subproducers. Creating these internal products below:
+        if self.call != None:
+            for p in self.producers:
```
Comment (harrypuuter) [inline code_generation/producer.py:107]: `for producer in self.producers`

**Review by harrypuuter — CHANGES_REQUESTED** [review]

**Review by harrypuuter — APPROVED** [review]

**Pull request merged**